### PR TITLE
CSR Approver

### LINKED ISSUE #83: Embed kubelet-rubber-stampr or something similar
We need to have kubelet to get a properly signed serving cert, otherwise configuring adjacent components such as metric server will get messy as by default kebelet is running with self-signed serving cert:
```
Server certificate
subject=CN = controller0@1597052929

issuer=CN = controller0-ca@1597052929
```

### PR DESCRIPTION
**Pull request by jasmingacic**
Fixes #83 

This PR also has a fix for konnectivity yaml which was malformed.

### CODE REVIEW

File: pkg/component/controller/csrapprover.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+	}
+	return recognizers
+}
+
+func (a *CSRApprover) isNodeServingCert(csr *v1.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{"system:nodes"}, x509cr.Subject.Organization) {
+		a.L.Warningf("Org does not match: %s", x509cr.Subject.Organization)
+		return false
+	}
+	if (len(x509cr.DNSNames) < 1) || (len(x509cr.IPAddresses) < 1) {
```
Comment (jnummelin) [inline pkg/component/controller/csrapprover.go:201]:
```suggestion
	if (len(x509cr.DNSNames) < 1) && (len(x509cr.IPAddresses) < 1) {
```

File: pkg/component/controller/csrapprover.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+				a.L.Info("endpoint reconciler done")
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (a *CSRApprover) approveCSR() error {
```
Comment (jnummelin) [inline pkg/component/controller/csrapprover.go:103]: Maybe add a note that this is adapted from rubber stamper

File: pkg/component/controller/csrapprover.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return csr, nil
+}
+
+func appendApprovalCondition(csr *v1.CertificateSigningRequest, message string) {
+	csr.Status.Conditions = append(csr.Status.Conditions, v1.CertificateSigningRequestCondition{
+		Type:    v1.CertificateApproved,
+		Reason:  "Autoapproved by K0S CSRApprover",
```
Comment (jnummelin) [inline pkg/component/controller/csrapprover.go:268]:
```suggestion
		Reason:  "Autoapproved by k0s CSRApprover",
```

**Review by jnummelin — COMMENTED** [review]
We definitely need some unit tests for the CSR approver parts as it's pretty critical stuff to get right

File: pkg/component/controller/csrapprover_test.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+)
+
+func TestBasicCRSApprover(t *testing.T) {
+	fakeFactory := testutil.NewFakeClientFactory()
+
+	client, err := fakeFactory.GetClient()
+	assert.NoError(t, err)
+	_, err = client.CoreV1().Endpoints("default").Get(context.TODO(), "kubernetes", v1.GetOptions{})
+	// The reconciler should not make any modification as we're not the leader so the endpoint should not get created
+	assert.Error(t, err)
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply pkg/component/controller/csrapprover_test.go]: Indeed

**Review by trawler — CHANGES_REQUESTED** [review]

File: pkg/component/controller/apiserver.go [hunk 5]
```diff
@@ -107,6 +107,7 @@ func (a *APIServer) Run() error {
 		"insecure-port":                    "0",
 		"profiling":                        "false",
 		"v":                                a.LogLevel,
+		"kubelet-certificate-authority":    path.Join(a.K0sVars.CertRootDir, "ca.crt"),
```
[the remark this answers is not in the document]
Comment (jasmingacic) [reply pkg/component/controller/apiserver.go:110]: The rest of the file uses only path.Join. I suppose then the entire file should be change

File: pkg/component/controller/csrapprover.go [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+var kubeletServerUsages = []v1.KeyUsage{
+	v1.UsageKeyEncipherment,
+	v1.UsageDigitalSignature,
+	v1.UsageServerAuth,
+}
```
[the remark this answers is not in the document]
Comment (jnummelin) [reply pkg/component/controller/csrapprover.go:45]: with the cert rotation enabled for both client & serving certs, kubelet will do CSRs for both separately. thus in the CSR approver we're only interested in the serving certs and safely ignore the possible client auth certs.

File: pkg/component/controller/apiserver.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jnummelin) [reply pkg/component/controller/apiserver.go:110]: Maybe keep this PR using `path.Join` as it's conformant with the rest of the usages in this file. Then a separate PR to refactor it to use `filepath.Join`

**Comment by jnummelin** [comment]
Triggered network conformance at: https://github.com/k0sproject/k0s/actions/runs/578058009

**Review by jnummelin — APPROVED** [review]

**Pull request merged**